src/compiler/checker.ts

```diff
@@ -5667,6 +5752,191 @@ namespace ts {
+                        members = undefined;
+                        stringIndexInfo = undefined;
+                        numberIndexInfo = undefined;
+                    }
+                    spreads.push(getTypeFromTypeNode((member as SpreadTypeElement).type));
+                }
+                else if (member.kind === SyntaxKind.IndexSignature) {
+                    const index = member as IndexSignatureDeclaration;
+                    if (index.parameters.length === 1) {
+                        const parameter = index.parameters[0];
+                        if (parameter && parameter.type) {
+                            const indexInfo = createIndexInfo(index.type ? getTypeFromTypeNode(index.type) : anyType,
+                                                              (getModifierFlags(index) & ModifierFlags.Readonly) !== 0, index);
+                            if (parameter.type.kind === SyntaxKind.StringKeyword) {
+                                stringIndexInfo = indexInfo;
+                            }
+                            else {
+                                numberIndexInfo = indexInfo;
+                            }
+                        }
+                    }
+                }
+                else if (member.kind !== SyntaxKind.CallSignature && member.kind !== SyntaxKind.ConstructSignature) {
+                    // note that spread types don't include call and construct signatures
+                    const flags = SymbolFlags.Property | SymbolFlags.Transient | (member.questionToken ? SymbolFlags.Optional : 0);
+                    const text = getTextOfPropertyName(member.name);
+                    const symbol = <TransientSymbol>createSymbol(flags, text);
+                    symbol.declarations = [member];
+                    symbol.valueDeclaration = member;
+                    symbol.type = getTypeFromTypeNodeNoAlias((member as IndexSignatureDeclaration | PropertySignature | MethodSignature).type);
+                    if (!members) {
+                        members = createMap<Symbol>();
+                    }
+                    members[symbol.name] = symbol;
+                }
+            }
+            if (members || stringIndexInfo || numberIndexInfo) {
+                spreads.push(createAnonymousType(node.symbol, members || emptySymbols, emptyArray, emptyArray, stringIndexInfo, numberIndexInfo));
+            }
+            return getSpreadType(spreads, node.symbol, aliasSymbol, aliasTypeArguments);
+        }
+
+        function getSpreadType(types: Type[], symbol: Symbol, aliasSymbol?: Symbol, aliasTypeArguments?: Type[]): Type {
+            if (types.length === 0) {
+                return emptyObjectType;
+            }
+            const id = getTypeListId(types);
+            if (id in spreadTypes) {
+                return spreadTypes[id];
+            }
+            let right = types.pop();
+            if (right.flags & TypeFlags.StringLike) {

```

Comment:
@sandersn Does that mean that if the right type is a string literal type, rather than a generic `string`, wouldn't it be more accurate to expand to the full, correct, object then? (For example, `type NSEW= {..."NSEW"};` can be expanded to `{0: "N", 1:"S", 2:"E", 3:"W"}` rather than `{[index: number]: string}`?)